kubelet: monitor plugin connections with application-level pings

bart0sh opened the pull request:
#### What type of PR is this?

/kind bug
/king feature

#### What this PR does / why we need it:

This PR implements a connection monitor for the Kubelet plugin watcher. The connection monitor periodically sends GetInfo requests to all registered plugins. If plugin stops responding to these requests it's automatically unregistered.
This implementation is used for Dynamic Resource Allocation (DRA), Container Storage Interface (CSI), and Device plugins.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/128696
Ref https://github.com/kubernetes/kubernetes/issues/127457

#### Does this PR introduce a user-facing change?

```release-note
Implemented a connection monitor for the Kubelet plugins. The connection monitor periodically sends GetInfo requests to all registered plugins. If a plugin stops responding to these requests, it is automatically unregistered. This implementation is used for Dynamic Resource Allocation (DRA), Container Storage Interface (CSI), and Device plugins.
```

bart0sh committed:
example kubelet plugin: unlink socket optionally

Added `unlinkSocket` attribute to the example plugin. This allows
to simulate stale sockets left on the filesystem after plugin crash.

bart0sh committed:
plugin watcher: add asw parameter

Added a new parameter to pass asw(Actual state of the world) to the
watcher. It will be used in the next commit to monitor plugin
connection.

bart0sh committed:
plugin_watcher: implement connection monitoring

bart0sh commented:
/sig node

bart0sh commented:
/cc @pohly @SergeyKanzhelev @swatisehgal @ffromani

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+// monitorPluginConnection monitors GRPC connection to the plugin
+// by periodically sending GetInfo RPC to the plugin
```
Why did you pick this approach instead of https://github.com/kubernetes-csi/csi-lib-utils/blob/e0fa89db2e21091a8d0ce49609ebef3ff8533b58/connection/connection.go#L203-L214?

This solution here can trigger for "server is still running and only unresponsive for a while" - I am not sure whether that's a pro or con.

pohly replied:
There's also https://grpc.io/docs/guides/keepalive/ but that says "It’s not required for service owners to support keepalive" where in this case "service owner" is the plugin.

pohly replied:
> As far as I understood it works only on connect. To monitor connected plugin we'd need to do something else.

No, it is meant to detect disconnects. That gRPC connects for a second time tells the custom dialer that the initial connection must have been lost.

The assumption is that gRPC reconnects. If it doesn't, then the connection loss is not detected. I don't remember whether that assumption was justified and whether it still works. I was hoping that you had tried that and found it to work resp. not work.

> I thought this is a pro as my idea was not only fix the issue with stalled sockets but to make registration reacting on plugin unresponsiveness.

What if the plugin is unresponsive long enough to get unregistered and becomes responsive again later?

I think it would be useful if you documented your changes in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md, including the rationale for those changes. Then it can be reviewed whether the code really works as intended and it might highlight semantic changes better than the diff alone can.

bart0sh replied:
> The assumption is that gRPC reconnects. If it doesn't, then the connection loss is not detected. I don't remember whether that assumption was justified and whether it still works. I was hoping that you had tried that and found it to work resp. not work.

When I was trying to use [WaitForStateChange](https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange) I didn't see any reconnection attempts when plugin (I used [example plugin](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin.go) + explicit listener close) stops serving. I can try your approach as well if you think it still makes sense. I'm happy with using RPC as it detects more cases, not just disconnects. Just let me know if you want me to try it.

> What if the plugin is unresponsive long enough to get unregistered and becomes responsive again later?

I can modify the implementation to catch this by trying RPC on unregistered or previously registered sockets and re-register the plugin. It will also help with use cases, mentioned in the https://github.com/kubernetes/kubernetes/pull/127821

pohly replied:
> When I was trying to use [WaitForStateChange](https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange) I didn't see any reconnection attempts when plugin (I used [example plugin](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin.go) + explicit listener close) stops serving. I can try your approach as well if you think it still makes sense.

You observation is correct, no need to try this. I checked the unit tests for the code I pointed to and they indeed rely on actually trying to use the connection. On an idle connection, gRPC doesn't try to reconnect. This probably could be fixed by enabling https://grpc.io/docs/guides/keepalive, but as discussed earlier, we cannot rely on that to be supported by plugins.

I think a good next step is to document the change in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md and then get some more opinions about the right way forward.

bart0sh replied:
Documented. Thanks for pointing out!

pohly commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// monitorPluginConnection monitors GRPC connection to the plugin
+// by periodically sending GetInfo RPC to the plugin.
```
See https://github.com/kubernetes/kubernetes/pull/129799#discussion_r1928681112.

bart0sh replied:
Can you elaborate? [I answered that comment](https://github.com/kubernetes/kubernetes/pull/129799#discussion_r1928856544), I believe.

pohly replied:
I'm just cross-referencing that thread here because it's still active and applies to this code here, but doesn't show up anymore when looking at a file diff.

bart0sh committed:
plugin_watcher: update documentation

Added a description of the plugin monitoring to the
pluginwatcher documentation.

Added DRA to the list of supported plugins.

bart0sh committed:
pluginmanager: aws: add PluginExists method

bart0sh committed:
example_plugin: add APIs to simulate stuckness

bart0sh commented:
/retest

SergeyKanzhelev commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,228 @@
+							m.connMap.Store(plugin.SocketPath, nil)
+							go m.monitorPluginConnection(plugin)
+						}
+					}
```
what does it mean if `PluginExists` returned false? Does it mean it will be registered shortly? Do we need to do anything here?

bart0sh replied:
It means that it's already registered as it present in both desired and actual state of world(dsw and asw). We should start monitoring it unless it's already monitored.

SergeyKanzhelev commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+// start starts the plugin connection monitor.
+func (m *pluginConnectionMonitor) start() {
```
with the existing of this function, can we trust that if `GetInfo` successful, we do not need to monitor the health of an actual plugin socket connection? Or for each plugin there will be two of those health loops?

bart0sh replied:
This function periodically discovers sockets that are not monitored and starts monitoring go-routine `monitorPluginConnection` for them one go-routine per connection, so there is only one loop per socket/connection.

SergeyKanzhelev commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,228 @@
+						if _, exists := m.connMap.Load(plugin.SocketPath); !exists {
+							m.connMap.Store(plugin.SocketPath, nil)
+							go m.monitorPluginConnection(plugin)
+						}
```
what would it mean if it is no in the `connMap`? Just trying to understand what are the else conditions there and whether we need to report them at least in log as an inconsistency?

bart0sh replied:
It would mean that this socket is not monitored yet. We don't need to report them as this code checks plugins every  `pluginListInterval` seconds and starts monitoring go-routine only once, when socket is not monitored (yet).

SergeyKanzhelev replied:
I am a bit worried in this and other `else` case that we will have some plugins "stuck" in registration somehow and we will not get any visibility into it. Logging from here will at least tell us that something is wrong

bart0sh replied:
Thanks for pointing out! Added more logging.

SergeyKanzhelev commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,228 @@
+	ctx, cancel := context.WithTimeout(context.Background(), getInfoTimeout)
+	defer cancel()
+
+	_, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
```
so what if the new info doesn't match old info? For example, plugin decided to start on a different socket. Should we re-register everything?

bart0sh replied:
If plugin decides to start on a different socket it will be registered again and a new monitoring go-routine will be run to monitor it. GetInfo here used as an application level ping, just to check that the plugin is alive and replies to the requests.

SergeyKanzhelev replied:
plugin starts on a same socket, but returns a new version of a plugin or a different endpoint for the plugin. We either need to use some other (new) method for health checks. Or respect changes returned from this method.

bart0sh replied:
Thank you for the review!

Added the code to update plugin attributes from the successful `GetInfo` RPC.

SergeyKanzhelev commented:
/sig storage

(to check the effect on the csi plugins)

SergeyKanzhelev commented:
I really like that this will add a retry on failed registration. So plugins will never disappear in limbo after one timeout from `GetInfo()`. However for ongoing monitoring - will we need this AND the monitoring fro the actual plugin socket? Can we jsut say that DRA MUST implement keep alive on the plugin socket?

bart0sh commented:
> will we need this AND the monitoring fro the actual plugin socket

In my opinion we'll either need both or can consider using one socket for everything as was proposed in https://github.com/kubernetes/kubernetes/pull/129809.

pohly commented:
I think it's okay to monitor only the registration socket. If the same process listens on both, then it seems very likely that "process is gone" or "process is stuck and doesn't process gRPC" occurs for both. If a process handles the registration for some other (CSI node-driver-registrar), then that process needs to monitor it's delegate (node-driver-registrar does or at least tries to) and will go away together with it.

The main purpose IMHO is presence detection, not health.

SergeyKanzhelev commented:
> The main purpose IMHO is presence detection, not health.

So the health monitoring will be a responsibility of a DRA plugin itself? I think it would be great to write it down as a best practice to DRA plugin authors

bart0sh committed:
pluginwatcher: implement monitoring stuck plugins

Added periodic checks for socket files in the plugins directory that
don't belong to any of the registered plugins.
If they respond to GetInfo RPC calls, connection manager re-registers
them.

tbs-test-stuck

bart0sh committed:
pluginwatcher: update plugin info on successful GetInfo RPC

bart0sh commented:
/retest

bart0sh commented:
/retest

bart0sh commented:
/retest

jsafrane commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -1,9 +1,13 @@
+directory returned by kubelet.getPluginsDir(). We will refer to this directory
+as PluginsDir.
+
+Additionally, it periodically sends Registration.GetInfo gRPC requests to the
+Unix sockets found in this directory to monitor plugin health.
```
A node may end up in a situation where the CSI driver pod looks working, but kubelet does thinks it's unhealthy. How can user figure that?

* Should we provide clear guidelines for the device / DRA / CSI plugins to monitor their end of the socket and do something when it gets closed?
* Is there a way how to recover from this situation? For example, restarting the CSI driver Pod could help, however, I don't see how kubelet could do it (it does not know what Pod created the socket).

pohly replied:
> Pods should be failing to run I guess when the driver is un-registered.

Isn't that impossible? The question here is about the situation where a plugin gets unregistered because it was failing to reply via gRPC. Not only is there no code in this PR which tries to tell the plugin that it got unregistered (or is there? - I haven't checked again), doing so at a time when the gRPC was unresponsive would probably not work.

bart0sh replied:
> Isn't that impossible? 

It's impossible to call an unregistered plugin if you're asking about it.

pohly replied:
So the question "Is there a way how to recover from this situation?" is still open because "Pods should be failing to run I guess when the driver is un-registered. " doesn't answer it, right?

bart0sh replied:
Nope. I can see one way to recover from this situation. This is what I described above - if pods are not running, cluster admin should be able to investigate it and restart the driver.

pohly replied:
The question was "A node may end up in a situation where the CSI driver pod looks working, but kubelet thinks it's unhealthy. ... Is there a way how to recover from this situation?".

In this case, the pod is running as far as the kubelet is concerned when monitoring the container, it's just not responding via gRPC long enough such that this new check unregisters the driver.

Will the node recover without manual intervention? If not, can the code be improved such that it does?

bart0sh replied:
> Will the node recover without manual intervention?

Yes, if driver starts to respond to GetInfo requests again, otherwise not.

> can the code be improved such that it does?

At least I don't see how to do it without changing registration API.

bart0sh replied:
Yes, there is a test `TestPluginStuckThenUnstuck` for this use case in this PR. I'm also working on e2e_node tests.

jsafrane replied:
Ack. I wrote https://github.com/kubernetes-csi/node-driver-registrar/issues/484 to check the CSI registrar that it behaves correctly when kubelet closes the socket.

jsafrane commented:
lgtm on the idea from storage POV. I did not review the code thoroughly enough.

bart0sh commented:
/cc @lauralorenz

bart0sh committed:
DRA: e2e_node: simulate plugin stuckness

Stuckness simulated by locking GetInfo call that is used
by Kubelet to monitor plugin connection.

bart0sh committed:
DRA: e2e_node: track plugin connections

bart0sh committed:
DRA test driver: add ResetRegistrationStatus method

bart0sh committed:
DRA: e2e_node: test monitoring plugin connections

Added new test case that simulates DRA plugin stuckness and expects
that plugin connection is dropped by Kubelet.
It also checks that plugin is re-registered after it becomes unstuck.

bart0sh commented:
@SergeyKanzhelev This PR is waiting for SIG-Node review and approval. SIG-storage lgtm-ed the idea. E2E and unit tests test have been added. PTAL.

bart0sh commented:
/assign @SergeyKanzhelev 
for a second review round

bart0sh commented:
/retest

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test /test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/retest

bart0sh commented:
@ffromani @pohly Can you review this PR again? I'm still hoping to get it merged before the code freeze.
CI failures don't seem to be related to this PR from the first look, but I'll continue looking.

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
@pohly May be. I'm trying to figure this out. So far I can only found these suspicious lines in the logs:
```
I0313 10:46:33.098042 68804 dump.go:53] At 2025-03-13 10:41:31 +0000 UTC - event for tester-1: {default-scheduler } FailedScheduling: 0/4 nodes are available: 1 node(s) had untolerated taint {node-role.kubernetes.io/control-plane: }, 3 cannot allocate all claims. still not schedulable, preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
I0313 10:46:33.098057 68804 dump.go:53] At 2025-03-13 10:41:35 +0000 UTC - event for tester-1: {default-scheduler } FailedScheduling: running "DynamicResources" filter plugin: pool network-attached is invalid: duplicate device name dev-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000
```

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-kind-dra

pohly commented on pkg/kubelet/pluginmanager/plugin_manager.go:
```diff
@@ -109,7 +109,7 @@ var _ PluginManager = &pluginManager{}
 	defer runtime.HandleCrash()
 
-	if err := pm.desiredStateOfWorldPopulator.Start(stopCh); err != nil {
+	if err := pm.desiredStateOfWorldPopulator.Start(stopCh, 2); err != nil {
```
Nit: why 2? What does it mean?

A named constant or a comment can answer that.

pohly commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -84,11 +88,13 @@ var _ ActualStateOfWorld = &actualStateOfWorld{}
+	UUID              types.UID
+	Handler           PluginHandler
+	Name              string
+	Endpoint          string
```
Can you document the difference between `SocketPath` and `Endpoint`?

I think I can guess it, but it helps to be more explicit.

bart0sh replied:
done

pohly commented on test/e2e/dra/test-driver/app/kubeletplugin.go:
```diff
@@ -555,3 +577,56 @@ func (ex *ExamplePlugin) CountCalls(methodSuffix string) int {
+}
+
+// grpc.stats.Handler interface method.
+// HandleConn is called when a GRPC connection starts and ends.
```
Is this mechanism reliable, in particular the "connection ends" part?

If yes, do we need the application level pings at all?

If no, then `h.connected.Store(false)` cannot be relied upon for testing.

bart0sh replied:
It's reliable on the GRPC server side. At least I didn't see it being unreliable. I'm not sure we can use it on the Kubelet side, at least without changing pluginmanager connections to be persistent. Current implementation closes them.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -78,6 +78,10 @@ const (
+
+	// pluginMonitorTimeout is the duration after which Kubelet closes the connection to the plugin
+	// if it doesn't respond to GetInfo GRPC calls.
+	pluginMonitorTimeout = time.Minute
```
Can this be a constant exported by a kubelet API package? Otherwise this might get out of sync with the implementation.

bart0sh replied:
This would be an awkward expression, which would be hard to understand. Something like this: dialTimeout + maxFailures * connectionCheckInterval + <some delay on top of it to ensure the test waited enough>

pohly replied:
If it's an awkward expression, then how can the test be sure that one minute is that duration? Or is the an estimate?

bart0sh replied:
It's an estimate based on that awkward expression and then proven by running the test many times.

pohly replied:
Then let's document it like that:

*"pluginMonitorTimeout is the duration after which Kubelet is expected to have closed the connection to the plugin if it doesn't get a response to the GetInfo gRPC calls. This is an upper bound, typically the kubelet is faster."*

bart0sh replied:
Thank you for the suggestion! Done.

bart0sh commented:
@SergeyKanzhelev @ffromani @swatisehgal @pohly PTAL. I'm hoping to get it merged before the code freeze with your help.

bart0sh commented:
/assign @ffromani @swatisehgal @pohly

swatisehgal commented on test/e2e_node/dra_test.go:
```diff
@@ -121,6 +125,24 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 			gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(pluginRegistrationTimeout).Should(testdriver.BeRegistered)
 		})
 
+		ginkgo.It("must unregister stuck plugin and re-register it when unstuck", func(ctx context.Context) {
```
I see that we have e2e tests for DRA, do we already have test coverage for monitoring plugin connections in case of device plugins and CSI Plugins?

bart0sh replied:
I haven't implemented any tests for Device Plugins and CSI.
@jsafrane at least [looked at this implementation](https://github.com/kubernetes/kubernetes/pull/129799#discussion_r1951142747) from the CSI/SIG-Storage point of view and accepted it.
And now you're looking :)

If you think it's absolutely required, I can implement either unit or e2e tests for Device Plugins, but I believe they're similar to DRA in this regard when using `pluginmanager` registration. Most of them are using old registration approach and should not be influenced by this PR.

swatisehgal reviewed (commented):
Disclosure (didn't go in too much detail as I am lacking context from DRA perspective)

Just a thought (perhaps for later) : Exposing metrics to capture plugin connections  (e.g. a GuageVector for exposing kubelet_plugin_connections  per plugin_name) would be a nice complement to this work and would provide aid to more observability  and help in the reliability effort https://github.com/kubernetes/kubernetes/issues/127457

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/monitor.go:
```diff
@@ -0,0 +1,220 @@
+	// It is used to monitor sockets that are not belong to registered plugins.
+	// As soon as the plugin becomes registered, the name is updated with the actual plugin name.
+	UndefinedPluginName = "Undefined"
+)
```
Well, the idea was not to make them too long as we don't want disconnects to stay not noticed for a long time, but not too short to not load the Kubelet too much. Then I adjusted the values a little bit based on tesing (both unit and e2e).

rphillips replied:
I worry the GetInfo() call is too heavy of a call. A device plugin might have to enumerate many devices to aggregate a bunch of information. Have we considered a light-weight Healthz style API call?

bart0sh replied:
@rphillips GetInfo() returns only general plugin info: type, name, path to the plugin unix socket and list of supported versions. I thought it's quite light to use as a plugin liveness check. Introducing new GRPC API would be harder and that wouldn't work with the existing plugins. 
Does this make sense?

rphillips replied:
Yes, thank you for the clarification.

pohly reviewed (commented):
Just some nits about comments.

Do we know who is going to do the final review and approval?

bart0sh commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -84,11 +88,18 @@ var _ ActualStateOfWorld = &actualStateOfWorld{}
 
 // PluginInfo holds information of a plugin
 type PluginInfo struct {
+	// SocketPath is a path to a unix socket file that the plugin is listening on
+	// for GRPC requests. Usually it's used for plugin registration.
```
Not only that. If GetInfo returns this socket or empty endpoint, SocketPath will be used as an endpoint. I tried to explain it below, describing what endpoint is.

pohly replied:
"The plugin may also provide other services via the same socket. SocketPath and Endpoint may be identical."

The key point is that for kubelet, this path is where it finds the registration service.

bart0sh replied:
Thank you for review! Applied your suggestions.

bart0sh commented:
@pohly 
> Do we know who is going to do the final review and approval?

I've presented this PR and asked for review and approval on the yesterday's sig-node meeting. @SergeyKanzhelev promised to look at it.
It was a pretty interesting discussion around this topic and SIG-Node leads were expressing certain concerns regarding the approach. One point was that if DRA plugin stops responding to the GetInfo requests Kubelet should remove its ResourceSlices objects. Another was that there are many layers of connection monitoring and we should think about reducing this number. There was also suggestion to look at the impact this approach does to the Kubelet in terms of performance.

bart0sh commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -84,11 +88,18 @@ var _ ActualStateOfWorld = &actualStateOfWorld{}
 	Handler    PluginHandler
 	Name       string
+	// Endpoint is the endpoint of the plugin provided by the plugin in the GetInfo response.
+	// If endpoint is not provided it will be set to the socket path byt the operation generator
```
I only saw this [in the code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go#L106) so far. I'll try to find the KEP if it exists and read it.

pohly replied:
This looks like an implementation detail. That's worth documenting here, but I wonder how useful that behavior is when it's not part of the contract between the kubelet and its plugins.

bart0sh replied:
I didn't find the KEP and not sure if it exists, but found this description in [the plugin registration API folder](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1/api.proto#L26-L31:


	// Optional endpoint location. If found set by Kubelet component,
	// Kubelet component will use this endpoint for specific requests.
	// This allows the plugin to register using one endpoint and possibly use
	// a different socket for control operations. CSI uses this model to delegate
	// its registration external from the plugin.
	string endpoint = 3;

bart0sh commented:
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-kind-dra-all
/test pull-kubernetes-kind-dra

bart0sh commented:
/retest

bart0sh commented:
/hold
I'm going to implement this differently on the DRA manager side. Will do it in a separate PR.

bart0sh commented:
Closing in favour of #131073
